assignment/doxygen/man/man3/Matrix.3
.TH "Matrix" 3 "Sun Dec 10 2017" "Heat conduction equation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Matrix \- A matrix class for data storage of a 2D array of doubles 
.br
 The implementation is derived from the standard container vector std::vector 
.br
 We use private inheritance to base our vector upon the library version whilst  usto expose only those base class functions we wish to use - in this  the array access operator []\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <matrix\&.h>\fP
.PP
Inherits vector< std::vector< double > >\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMatrix\fP ()"
.br
.RI "\fIDefault constructor\&. \fP"
.ti -1c
.RI "\fBMatrix\fP (int Nrows, int Ncols)"
.br
.RI "\fIAlternate constructor\&. \fP"
.ti -1c
.RI "\fBMatrix\fP (const \fBMatrix\fP &m)"
.br
.RI "\fICopy constructor\&. \fP"
.ti -1c
.RI "int \fBgetNrows\fP () const "
.br
.RI "\fINormal public get method\&. \fP"
.ti -1c
.RI "int \fBgetNcols\fP () const "
.br
.RI "\fINormal public get method\&. \fP"
.ti -1c
.RI "void \fBset_row\fP (int index, \fBVector\fP v)"
.br
.RI "\fINormal public set method\&. \fP"
.ti -1c
.RI "\fBMatrix\fP & \fBoperator=\fP (const \fBMatrix\fP &m)"
.br
.RI "\fIOverloaded assignment operator\&. \fP"
.ti -1c
.RI "bool \fBoperator==\fP (const \fBMatrix\fP &m) const "
.br
.RI "\fIOverloaded comparison operator returns true or false depending on whether the matrices are the same or not\&. \fP"
.ti -1c
.RI "double \fBone_norm\fP () const "
.br
.RI "\fINormal public method that returns a double\&. \fP"
.ti -1c
.RI "double \fBtwo_norm\fP () const "
.br
.RI "\fINormal public method that returns a double\&. \fP"
.ti -1c
.RI "double \fBuniform_norm\fP () const "
.br
.RI "\fINormal public method that returns a double\&. \fP"
.ti -1c
.RI "\fBMatrix\fP \fBoperator*\fP (const \fBMatrix\fP &a) const "
.br
.RI "\fIOverloaded *operator that returns a \fBMatrix\fP\&. \fP"
.ti -1c
.RI "\fBVector\fP \fBoperator*\fP (const \fBVector\fP &v) const "
.br
.RI "\fIOverloaded *operator that returns a \fBVector\fP\&. \fP"
.ti -1c
.RI "\fBMatrix\fP \fBtranspose\fP () const "
.br
.RI "\fIpublic method that returns the transpose of the matrix\&. \fP"
.ti -1c
.RI "\fBMatrix\fP \fBmult\fP (const \fBMatrix\fP &a) const "
.br
.in -1c
.SS "Private Types"

.in +1c
.ti -1c
.RI "typedef std::vector< std::vector< double > > \fBvec\fP"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "std::istream & \fBoperator>>\fP (std::istream &is, \fBMatrix\fP &m)"
.br
.RI "\fIOverloaded istream >> operator\&. \fP"
.ti -1c
.RI "std::ostream & \fBoperator<<\fP (std::ostream &os, const \fBMatrix\fP &m)"
.br
.RI "\fIOverloaded ostream << operator\&. \fP"
.ti -1c
.RI "std::ifstream & \fBoperator>>\fP (std::ifstream &ifs, \fBMatrix\fP &m)"
.br
.RI "\fIOverloaded ifstream >> operator\&. \fP"
.ti -1c
.RI "std::ofstream & \fBoperator<<\fP (std::ofstream &ofs, const \fBMatrix\fP &m)"
.br
.RI "\fIOverloaded ofstream << operator\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
A matrix class for data storage of a 2D array of doubles 
.br
 The implementation is derived from the standard container vector std::vector 
.br
 We use private inheritance to base our vector upon the library version whilst  usto expose only those base class functions we wish to use - in this  the array access operator []\&. 

The \fBMatrix\fP class provides: 
.br
-basic constructors for creating a matrix object from other matrix object,  by creating empty matrix of a given size, 
.br
-input and oput operation via >> and << operators using keyboard or file 
.br
-basic operations like access via [] operator, assignment and comparision 
.SH "Member Typedef Documentation"
.PP 
.SS "typedef std::vector<std::vector<double> > \fBMatrix::vec\fP\fC [private]\fP"

.SH "Constructor & Destructor Documentation"
.PP 
.SS "Matrix::Matrix ()"

.PP
Default constructor\&. Intialize an empty \fBMatrix\fP object 
.PP
\fBSee also:\fP
.RS 4
\fBMatrix(int Nrows, int Ncols)\fP 
.PP
\fBMatrix(const Matrix& m)\fP 
.RE
.PP

.SS "Matrix::Matrix (int Nrows, int Ncols)"

.PP
Alternate constructor\&. build a matrix Nrows by Ncols 
.PP
\fBSee also:\fP
.RS 4
\fBMatrix()\fP 
.PP
\fBMatrix(const Matrix& m)\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIinvalid_argument\fP ('matrix size negative or zero') 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fINrows\fP int\&. number of rows in matrix 
.br
\fINcols\fP int\&. number of columns in matrix 
.RE
.PP

.SS "Matrix::Matrix (const \fBMatrix\fP & m)"

.PP
Copy constructor\&. build a matrix from another matrix 
.PP
\fBSee also:\fP
.RS 4
\fBMatrix()\fP 
.PP
\fBMatrix(int Nrows, int Ncols)\fP 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIm\fP \fBMatrix\fP&\&. matrix to copy from 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "int Matrix::getNcols () const"

.PP
Normal public get method\&. get the number of columns 
.PP
\fBSee also:\fP
.RS 4
int \fBgetNrows()const\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
int\&. number of columns in matrix 
.RE
.PP

.SS "int Matrix::getNrows () const"

.PP
Normal public get method\&. get the number of rows 
.PP
\fBSee also:\fP
.RS 4
int \fBgetNcols()const\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
int\&. number of rows in matrix 
.RE
.PP

.SS "\fBMatrix\fP Matrix::mult (const \fBMatrix\fP & a) const"

.SS "double Matrix::one_norm () const"

.PP
Normal public method that returns a double\&. It returns L1 norm of matrix 
.PP
\fBSee also:\fP
.RS 4
\fBtwo_norm()const\fP 
.PP
\fBuniform_norm()const\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
double\&. matrix L1 norm 
.RE
.PP

.SS "\fBMatrix\fP Matrix::operator* (const \fBMatrix\fP & a) const"

.PP
Overloaded *operator that returns a \fBMatrix\fP\&. It Performs matrix by matrix multiplication\&. 
.PP
\fBSee also:\fP
.RS 4
\fBoperator*(const Matrix & a) const\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIout_of_range\fP ('Matrix access error') One or more of the matrix have a zero size 
.br
\fIstd::out_of_range\fP ('uncompatible matrix sizes') Number of columns in first matrix do not match number of columns in second matrix 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBMatrix\fP\&. matrix-matrix product 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIa\fP \fBMatrix\fP\&. matrix to multiply by 
.RE
.PP

.SS "\fBVector\fP Matrix::operator* (const \fBVector\fP & v) const"

.PP
Overloaded *operator that returns a \fBVector\fP\&. It Performs matrix by vector multiplication\&. 
.PP
\fBSee also:\fP
.RS 4
\fBoperator*(const Matrix & a)const\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIstd::out_of_range\fP ('Matrix access error') matrix has a zero size 
.br
\fIstd::out_of_range\fP ('Vector access error') vector has a zero size 
.br
\fIstd::out_of_range\fP ('uncompatible matrix-vector sizes') Number of columns in matrix do not match the vector size 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBVector\fP\&. matrix-vector product 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIv\fP \fBVector\fP\&. \fBVector\fP to multiply by 
.RE
.PP

.SS "\fBMatrix\fP & Matrix::operator= (const \fBMatrix\fP & m)"

.PP
Overloaded assignment operator\&. 
.PP
\fBSee also:\fP
.RS 4
\fBoperator==(const Matrix& m)const\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBMatrix\fP&\&. the matrix on the left of the assignment 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIm\fP \fBMatrix\fP&\&. \fBMatrix\fP to assign from 
.RE
.PP

.SS "bool Matrix::operator== (const \fBMatrix\fP & m) const"

.PP
Overloaded comparison operator returns true or false depending on whether the matrices are the same or not\&. 
.PP
\fBSee also:\fP
.RS 4
\fBoperator=(const Matrix& m)\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
bool\&. true or false 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIm\fP \fBMatrix\fP&\&. \fBMatrix\fP to compare to 
.RE
.PP

.SS "void Matrix::set_row (int index, \fBVector\fP v)"

.PP
Normal public set method\&. replace a row with a given vector 
.PP
\fBParameters:\fP
.RS 4
\fIindex\fP Index of row to mutate 
.br
\fIv\fP New vector 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIout_of_range\fP ('index out of range\&.\\n') 
.br
\fIout_of_range\fP ('vector size is different from matrix columns number\&.\\n') 
.RE
.PP

.SS "\fBMatrix\fP Matrix::transpose () const"

.PP
public method that returns the transpose of the matrix\&. It returns the transpose of matrix 
.PP
\fBReturns:\fP
.RS 4
\fBMatrix\fP\&. matrix transpose 
.RE
.PP

.SS "double Matrix::two_norm () const"

.PP
Normal public method that returns a double\&. It returns L2 norm of matrix 
.PP
\fBSee also:\fP
.RS 4
\fBone_norm()const\fP 
.PP
\fBuniform_norm()const\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
double\&. matrix L2 norm 
.RE
.PP

.SS "double Matrix::uniform_norm () const"

.PP
Normal public method that returns a double\&. It returns L_max norm of matrix 
.PP
\fBSee also:\fP
.RS 4
\fBone_norm()const\fP 
.PP
\fBtwo_norm()const\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
double\&. matrix L_max norm 
.RE
.PP

.SH "Friends And Related Function Documentation"
.PP 
.SS "std::ostream& operator<< (std::ostream & os, const \fBMatrix\fP & m)\fC [friend]\fP"

.PP
Overloaded ostream << operator\&. Display output if matrix has size user will be asked to input only matrix values if matrix was not initialized user can choose matrix size and input it values 
.PP
\fBSee also:\fP
.RS 4
\fBoperator>>(std::ifstream& ifs, Matrix& m)\fP 
.PP
\fBoperator>>(std::istream& is, Matrix& m)\fP 
.PP
\fBoperator<<(std::ostream& os, const Matrix& m)\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
std::ostream&\&. The ostream object 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIos\fP Display output stream 
.br
\fIm\fP \fBMatrix\fP to read from 
.RE
.PP

.SS "std::ofstream& operator<< (std::ofstream & ofs, const \fBMatrix\fP & m)\fC [friend]\fP"

.PP
Overloaded ofstream << operator\&. File output the file output operator is compatible with file input operator, ie\&. everything written can be read later\&. 
.PP
\fBSee also:\fP
.RS 4
\fBoperator>>(std::ifstream& ifs, Matrix& m)\fP 
.PP
\fBoperator<<(std::ofstream& ofs, const Matrix& m)\fP 
.PP
\fBoperator>>(std::istream& is, Matrix& m)\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIstd::invalid_argument\fP ('file read error - negative matrix size'); 
.RE
.PP
\fBReturns:\fP
.RS 4
std::ofstream&\&. The ofstream object 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIm\fP \fBMatrix\fP to read from 
.RE
.PP

.SS "std::istream& operator>> (std::istream & is, \fBMatrix\fP & m)\fC [friend]\fP"

.PP
Overloaded istream >> operator\&. Keyboard input if matrix has size user will be asked to input only matrix values if matrix was not initialized user can choose matrix size and input it values 
.PP
\fBSee also:\fP
.RS 4
\fBoperator<<(std::ofstream& ofs, const Matrix& m)\fP 
.PP
\fBoperator>>(std::istream& is, Matrix& m)\fP 
.PP
\fBoperator<<(std::ostream& os, const Matrix& m)\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIstd::invalid_argument\fP ('read error - negative matrix size'); 
.RE
.PP
\fBReturns:\fP
.RS 4
std::istream&\&. The istream object 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIis\fP Keyboard input stream 
.br
\fIm\fP \fBMatrix\fP to write into 
.RE
.PP

.SS "std::ifstream& operator>> (std::ifstream & ifs, \fBMatrix\fP & m)\fC [friend]\fP"

.PP
Overloaded ifstream >> operator\&. File input the file output operator is compatible with file input operator, ie\&. everything written can be read later\&. 
.PP
\fBSee also:\fP
.RS 4
\fBoperator>>(std::ifstream& ifs, Matrix& m)\fP 
.PP
\fBoperator<<(std::ofstream& ofs, const Matrix& m)\fP 
.PP
\fBoperator<<(std::ostream& os, const Matrix& m)\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
std::ifstream&\&. The ifstream object 
.RE
.PP

.PP
\fBParameters:\fP
.RS 4
\fIifs\fP Input file stream with opened matrix file 
.br
\fIm\fP \fBMatrix\fP to write into 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for Heat conduction equation from the source code\&.


reports/src/grid/matrix.cpp

#include "matrix.h"

// CONSTRUCTORS
/*=
 *Default constructor (empty matrix)
 */
Matrix::Matrix() : std::vector<std::vector<double> >()  {}

/*
 * Alternate constructor - creates a matrix with the given values
 */
Matrix::Matrix(int Nrows, int Ncols) : std::vector<std::vector<double> >()
{
    //check input
    if(Nrows < 0 || Ncols < 0) throw std::invalid_argument("matrix size negative");
    
	// set the size for the rows
	(*this).resize(Nrows);
	// set the size for the columns
	for (int i = 0; i < Nrows; i++) (*this)[i].resize(Ncols);

	// initialise the matrix to contain zero
	for (int i = 0; i < Nrows; i++)
		for (int j = 0; j < Ncols; j++) (*this)[i][j] = 0;
}


/*
 * Copy constructor
 */
Matrix::Matrix(const Matrix& m) : std::vector<std::vector<double> >()
{
	// set the size of the rows
	(*this).resize(m.size());
	// set the size of the columns
	std::size_t i;
	for (i = 0; i < m.size(); i++) (*this)[i].resize(m[0].size());

	// copy the elements
	for (int i = 0; i < m.getNrows(); i++)
		for (int j = 0; j < m.getNcols(); j++)
			(*this)[i][j] = m[i][j];
}

// ACCESSOR METHODS
/*
* accessor method - get the number of rows
*/
int Matrix :: getNrows() const
{
	return size();
}

/*
* accessor method - get the number of columns
*/
int Matrix::getNcols() const
{
	return (*this)[0].size();
}

// MUTATORS METHODS

/*
* mutator method - set new values to a row
*/

void Matrix::set_row(int index, Vector v) {
	int vector_size = v.getSize();

	//catches invalid arguments 
	if (index < 0 || index >= vector_size) throw std::out_of_range("index out of range.");
	if (vector_size != (*this).getNcols()) throw std::out_of_range("vector size is different from matrix columns number.");

	for (unsigned int i = 0; i < vector_size; i++) {
		(*this)[index][i] = v[i];
	}
}

// NORMS
/*
 * 1 norm
 */
double Matrix::one_norm() const
{
	int nrows = getNrows();
	int ncols = getNcols();
	Matrix transpose = this->transpose();
	Vector * sizes = new Vector(nrows);
	for (unsigned int index = 0; index < nrows; index++) {
		Vector * vector = (Vector*)&transpose[index];
		(*sizes)[index] = vector->one_norm();
	}
	return sizes->uniform_norm();
}

/*
 * 2 norm
 */
double Matrix::two_norm() const
{
	int nrows = getNrows();
	int ncols = getNcols();

	double result = 0.0;
	for (unsigned int i = 0; i < nrows; i++) {
		for (unsigned int j = 0; j < ncols; j++) {
			result += pow((*this)[i][j], 2);
		}
	}
	return sqrt(result);
}

/*
 * uniform (infinity) norm
 */
double Matrix::uniform_norm() const
{
	int nrows = getNrows();
	int ncols = getNcols();
	Vector * sizes = new Vector(nrows);
	for (unsigned int index = 0; index < nrows; index++) {
		Vector * vector = (Vector*)&((*this)[index]);
		(*sizes)[index] = vector->one_norm();
	}
	return sizes->uniform_norm();
}

// OVERLOADED OPERATORS
/*
* Operator= - assignment
*/
Matrix& Matrix::operator=(const Matrix& m)
{	if (this != &m) {
		(*this).resize(m.size());
		std::size_t i;
		std::size_t j;
		for (i = 0; i < m.size(); i++) (*this)[i].resize(m[0].size());

		for (i = 0; i<m.size(); i++)
			for (j = 0; j<m[0].size(); j++)
				(*this)[i][j] = m[i][j];
	}
	return *this;
}

/*
* Operator* multiplication of a matrix by a matrix
*/
Matrix Matrix::operator*(const Matrix& a) const {

	int nrows = getNrows();
	int ncols = getNcols();
	
	// catch invalid matrices
	if (nrows <= 0 || ncols <= 0) { throw std::out_of_range("Matrix access error"); }
	if (a.getNrows() <= 0 || a.getNcols() <= 0) { throw std::out_of_range("Matrix access error"); }

	//if the matrix sizes do not match
	if (ncols != a.getNrows()) throw std::out_of_range("matrix sizes do not match");

	//  matrix to store the result
	Matrix mmult = Matrix(getNrows(), a.getNcols());

	//matrix multiplication
	for (int i = 0;i<getNrows();i++) {
		for (int j = 0;j<a.getNcols();j++) {
			for (int k = 0;k<getNcols();k++) {
				mmult[i][j] += ((*this)[i][k] * a[k][j]);
			}
		}
	}
	return mmult;
}


/*
* Operator* multiplication of a matrix by a vector
*/
Vector Matrix::operator*(const Vector& v) const {
	int nrows = getNrows();
	int ncols = getNcols();

	// catch invalid matrix, vector
	if (nrows <= 0 || ncols <=0 ) { throw std::out_of_range("Matrix access error"); }
	if (v.getSize() <= 0) { throw std::out_of_range("Vector access error"); }

	//if the matrix sizes do not match
	if (ncols != v.getSize()) throw std::out_of_range("matrix sizes do not match");

	// matrix to store the multiplication
	Vector res(nrows);

	// perform the multiplication
	for (int i = 0;i<nrows;i++)
		for (int j = 0;j<ncols;j++) res[i] += ((*this)[i][j] * v[j]);

	// return the result
	return res;
}


/*
 * Operator== comparison function, returns true if the given matrices are the same
 */
bool Matrix::operator==(const Matrix& a) const {
	int nrows = getNrows();
	int ncols = getNcols();

	//if the sizes do not match return false*
	if ( (nrows != a.getNrows()) || (ncols != a.getNcols()) ) return false;

	//compare all of the elements
	for (int i=0;i<nrows;i++) {
		for (int j=0;j<ncols;j++) {
			if (fabs((*this)[i][j] - a[i][j]) > 1.e-07) { return false; }
		}
	}

	return true;
}


// OTHER METHODS
/*
 * Transpose of the matrix
 */
Matrix Matrix::transpose() const
{
	int nrows = getNrows();
	int ncols = getNcols();

	// matrix to stoore the transpose
	Matrix temp(ncols, nrows);

	for (int i=0; i < ncols; i++)
		for (int j=0; j < nrows; j++)
			temp[i][j] = (*this)[j][i];

	return temp;
}

Matrix Matrix::mult(const Matrix& m) const
{
	return (*this) * m;
}



// INPUT AND OUTPUT FUNCTIONS
/*
 * keyboard input , user friendly
 */
std::istream& operator>>(std::istream& is, Matrix& m) {
	
	int nrows, ncols;

	// test to see whether the matrix m is empty
	if (!m.getNrows()) {
		std::cout << "input the number of rows for the matrix" << std::endl;
		is >> nrows;
		std::cout << "input the number of cols for the matrix" << std::endl;
		is >> ncols;
		//check input 
		if(nrows < 0 || ncols < 0) throw std::invalid_argument("read error - negative matrix size");

		// prepare the matrix to hold n elements
		m = Matrix(nrows, ncols);
	}
    // input the elements
    std::cout << "input "<< m.getNrows() * m.getNcols() << " matrix elements" << std::endl;
	for (int i = 0; i < m.getNrows(); i++)
		for (int j=0; j< m.getNcols(); j++) is >> m[i][j];

    // return the stream object
    return is;
}

/* 
 * screen output, user friendly
 */
std::ostream& operator<<(std::ostream& os, const Matrix& m) {

	// test to see whether there are any elements
    if (m.getNrows() > 0) {
        os << "The matrix elements are" << std::endl;
        for (int i=0; i<m.getNrows();i++) {
			for (int j=0;j<m.getNcols();j++) {
				os << m[i][j]  << " ";
			}
			os << "\n";
		}
        os << std::endl;
    }
    else
    {
        os << "Matrix is empty." << std::endl;
    }
    return os;
}

/*
* file input - raw data, compatible with file writing operator
*/
std::ifstream& operator>>(std::ifstream& ifs, Matrix& m) {
	
	int nrows, ncols;

    // read size from the file
    ifs >> nrows; ifs>> ncols;
    //check input sanity
    if(nrows < 0 || ncols < 0) throw std::invalid_argument("file read error - negative matrix size");

    // prepare the vector to hold n elements
    m = Matrix(nrows, ncols);

    // input the elements
    for (int i=0; i<nrows; i++) 
		for (int j=0; j<ncols; j++) ifs >> m[i][j];

    // return the stream object
    return ifs;
}


/*
* file output - raw data, comaptible with file reading operator
*/
std::ofstream& operator<<(std::ofstream& ofs, const Matrix& m) {
    //put matrix rownumber in first line (even if it is zero)
    ofs << m.getNrows() << std::endl;
	//put matrix columnnumber in second line (even if it is zero)
	ofs << m.getNcols() << std::endl;
    //put data in third line (if size==zero nothing will be put)
    for (int i=0; i<m.getNrows(); i++) {
		for (int j=0; j<m.getNcols(); j++) ofs << m[i][j] <<  " ";
		ofs << std::endl;
	}
    return ofs;
}

